addons/product_margin/product_margin.py

```diff
@@ -71,68 +45,44 @@ def _product_margin(self, cr, uid, ids, field_names, arg, context=None):
                 where l.product_id = %s and i.state in %s and i.type IN %s and (i.date_invoice IS NULL or (i.date_invoice>=%s and i.date_invoice<=%s and i.company_id=%s))
                 """
             invoice_types = ('out_invoice', 'in_refund')
-            cr.execute(sqlstr, (val.id, states, invoice_types, date_from, date_to, company_id))
-            result = cr.fetchall()[0]
-            res[val.id]['sale_avg_price'] = result[0] and result[0] or 0.0
-            res[val.id]['sale_num_invoiced'] = result[1] and result[1] or 0.0
-            res[val.id]['turnover'] = result[2] and result[2] or 0.0
-            res[val.id]['sale_expected'] = result[3] and result[3] or 0.0
-            res[val.id]['sales_gap'] = res[val.id]['sale_expected']-res[val.id]['turnover']
-            prod_obj = self.pool.get("product.product")
-            ctx = context.copy()
-            ctx['force_company'] = company_id
-            prod = prod_obj.browse(cr, uid, val.id, context=ctx)
+            self.env.cr.execute(sqlstr, (margin.id, states, invoice_types, date_from, date_to, company_id))
+            result = self.env.cr.fetchall()[0]
+            margin.sale_avg_price = result[0] and result[0] or 0.0
+            margin.sale_num_invoiced = result[1] and result[1] or 0.0
+            margin.turnover = result[2] and result[2] or 0.0
+            margin.sale_expected = result[3] and result[3] or 0.0
+            margin.sales_gap = margin.sale_expected - margin.turnover
+            Product = self.env['product.product']
+            Prod = Product.with_context(force_company=company_id).browse(margin.id)
             invoice_types = ('in_invoice', 'out_refund')
-            cr.execute(sqlstr, (val.id, states, invoice_types, date_from, date_to, company_id))
-            result = cr.fetchall()[0]
-            res[val.id]['purchase_avg_price'] = result[0] and result[0] or 0.0
-            res[val.id]['purchase_num_invoiced'] = result[1] and result[1] or 0.0
-            res[val.id]['total_cost'] = result[2] and result[2] or 0.0
-            res[val.id]['normal_cost'] = prod.standard_price * res[val.id]['purchase_num_invoiced']
-            res[val.id]['purchase_gap'] = res[val.id]['normal_cost'] - res[val.id]['total_cost']
+            self.env.cr.execute(sqlstr, (margin.id, states, invoice_types, date_from, date_to, company_id))
+            result = self.env.cr.fetchall()[0]
+            margin.purchase_avg_price = result[0] and result[0] or 0.0
+            margin.purchase_num_invoiced = result[1] and result[1] or 0.0
+            margin.total_cost = result[2] and result[2] or 0.0
+            margin.normal_cost = Prod.standard_price * margin.purchase_num_invoiced
+            margin.purchase_gap = margin.normal_cost - margin.total_cost
 
```

Comment:
to much assign value. call write. can try one write method instead of assignment.